pravega-client-examples/src/test/java/io/pravega/utils/SetupUtils.java

```diff
@@ -0,0 +1,353 @@
+/*
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ */
+package io.pravega.utils;
+
+import com.google.common.base.Preconditions;
+import io.pravega.client.ClientConfig;
+import io.pravega.client.EventStreamClientFactory;
+import io.pravega.client.admin.StreamManager;
+import io.pravega.client.stream.ScalingPolicy;
+import io.pravega.client.stream.Stream;
+import io.pravega.client.stream.StreamConfiguration;
+import io.pravega.client.stream.impl.Controller;
+import io.pravega.client.stream.impl.ControllerImpl;
+import io.pravega.client.stream.impl.ControllerImplConfig;
+import io.pravega.common.concurrent.ExecutorServiceHelpers;
+import io.pravega.local.InProcPravegaCluster;
+import lombok.Cleanup;
+import lombok.Getter;
+import lombok.Setter;
+import org.apache.commons.lang3.RandomStringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.concurrent.NotThreadSafe;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.ServerSocket;
+import java.net.URI;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardCopyOption;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Utility functions for creating an in-process Pravega server or connecting to an external Pravega server.
+ */
+@NotThreadSafe
```

Comment:
This is sort of supprising given the class's name. Perhaps a different name should be used?

Reply:
I borrowed this entire class from https://github.com/fpj/eo-ingestion/blob/master/src/test/java/io/pravega/utils/SetupUtils.java and made minimal changes. This class appears to be based on https://github.com/pravega/pravega/blob/master/test/integration/src/main/java/io/pravega/test/integration/utils/SetupUtils.java. All of these are annotated with `@NotThreadSafe`.

Reply:
Maybe this https://github.com/pravega/pravega/blob/master/test/integration/src/test/java/io/pravega/test/integration/PravegaResource.java would would be a better model. It takes advantage of the `@Rule` syntax to allow for single initialization without imposing thread safety issues.